src/LevelZeroShim.hpp

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (c) 2015 - 2021, Intel Corporation
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in
+ *       the documentation and/or other materials provided with the
+ *       distribution.
+ *
+ *     * Neither the name of Intel Corporation nor the names of its
+ *       contributors may be used to endorse or promote products derived
+ *       from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY LOG OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef LEVELZEROSHIM_HPP_INCLUDE
+#define LEVELZEROSHIM_HPP_INCLUDE
+
+#include <vector>
+#include <string>
+#include <cstdint>
+
+#include "geopm_sched.h"
+#include "geopm_topo.h"
+
+namespace geopm
+{
+    class LevelZeroShim
+    {
+        public:
+            enum geopm_levelzero_domain_e {
+                GEOPM_LEVELZERO_DOMAIN_ALL = 0,
+                GEOPM_LEVELZERO_DOMAIN_COMPUTE = 1,
+                GEOPM_LEVELZERO_DOMAIN_MEMORY = 2,
+                GEOPM_LEVELZERO_DOMAIN_SIZE = 3
+            };
+
+            LevelZeroShim() = default;
+            virtual ~LevelZeroShim() = default;
+            /// @brief Number of accelerators on the platform.
+            /// @return Number of LevelZero accelerators.
+            virtual int num_accelerator(void) const = 0;
+
+            /// @brief Number of accelerator subdevices on the platform.
+            /// @return Number of LevelZero accelerator subdevices.
+            virtual int num_accelerator_subdevice(void) const = 0;
+
+            /// @brief Get the number of LevelZero frequency domains of a certain type
+            /// @param [in] domain The domain type being targeted
```

Comment:
```suggestion
            /// @param [in] domain The GEOPM domain type being targeted
            /// @param [in] domain_idx The GEOPM domain index being targeted            
            /// @param [in] l0_component The L0 component being targeted
```

If it makes sense to stick with device and device_idx here because the DevicePool is doing that translation, then so be it.

Reply:
Moved to the more explicit l0_domain to differentiate between GEOPM domains (domain) and level zero domains (l0_domain).  There are few instances of GEOPM domains in the shim (as the device pool handles GEOPM domain to Level Zero domain, device, and sub-device conversion), so this change was made in the device pool as well.